File: 프로그래머스/2/118667. 두 큐 합 같게 만들기/두 큐 합 같게 만들기.py
from collections import deque


def solution(queue1, queue2):
    v1 = sum(queue1)
    v2 = sum(queue2)
    
    div, mod = divmod(v1 + v2, 2)
    if mod:
        return -1

    cnt = 0
    q1 = deque(queue1)
    q2 = deque(queue2)

    while q1 and q2:
        if v1 == div:
            return cnt
        
        cnt += 1
        if v1 >= div:
            v1 -= q1.popleft()
            continue
        
        pop = q2.popleft()
        q1.append(pop)
        v1 += pop

    return -1

File: 프로그래머스/2/118667. 두 큐 합 같게 만들기/두 큐 합 같게 만들기.java
import java.util.*;

class Solution {
    
    public int solution(int[] queue1, int[] queue2) {
        long sum1 = 0, sum2 = 0;
        
        for (int n : queue1) sum1 += n;
        for (int n : queue2) sum2 += n;
        
        long total = sum1 + sum2;
        if (total % 2 != 0) return -1;
        
        long target = total / 2;
        
        Deque<Integer> q1 = new ArrayDeque<>();
        Deque<Integer> q2 = new ArrayDeque<>();
        
        for (int n : queue1) {
            q1.offer(n);
        }
        for (int n : queue2) {
            q2.offer(n);
        }
        
        int count = 0;
        int limit = (queue1.length + queue2.length) * 3;
        
        while (!q1.isEmpty() && !q2.isEmpty() && count <= limit) {
            if (sum1 == target) return count;
            
            if (sum1 > target) {
                int x = q1.poll();
                sum1 -= x;
                q2.offer(x);
            } else {
                int x = q2.poll();
                sum1 += x;
                q1.offer(x);
            }
            
            count++;
        }
        
        return -1;
    }
}